components/style/parser.rs

```diff
@@ -9,15 +9,32 @@ use selectors::parser::ParserContext as SelectorParserContext;
 use stylesheets::Origin;
 use url::Url;
 
+#[cfg(not(feature = "gecko"))]
+pub type ParserContextExtraData = ();
+
+#[cfg(feature = "gecko")]
+pub struct ParserContextExtraData {
+}
+
+#[cfg(not(feature = "gecko"))]
+pub fn dummy_extra_data() {}
+
+#[cfg(feature = "gecko")]
+pub fn dummy_extra_data() -> ParserContextExtraData {
+    ParserContextExtraData { }
+}
```

Comment:
Instead of exposing this everywhere, I think we should have a ParserContext::new and ParserContext::new_with_extra_data, where the former delegates to the latter and passes in a sensible default. Then we can make ::default() a static factory method on ParserContextExtraData.